swiftide-integrations/src/openai/mod.rs

```diff
@@ -211,9 +213,60 @@ impl<C: async_openai::config::Config + Default> GenericOpenAI<C> {
     }
 }
 
+pub fn openai_error_to_language_model_error(e: OpenAIError) -> LanguageModelError {
+    match e {
+        OpenAIError::ApiError(api_error) => {
+            // If the response is an ApiError, it could be a context length exceeded error
+            if api_error.code == Some("context_length_exceeded".to_string()) {
+                LanguageModelError::ContextLengthExceeded(OpenAIError::ApiError(api_error).into())
+            } else {
+                LanguageModelError::PermanentError(OpenAIError::ApiError(api_error).into())
+            }
+        }
+        OpenAIError::Reqwest(e) => {
+            if let Some(status) = e.status() {
+                // If the response code is 429 it could either be a TransientError or a
+                // PermanentError depending on the message, if it contains the word
+                // quota, it should be a PermanentError otherwise it should
+                // be a TransientError.
+                // If the response code is any other 4xx it should be a PermanentError.
+                if status.as_u16() == 429 && !e.to_string().contains("quota") {
+                    LanguageModelError::TransientError(e.into())
+                } else if status.is_client_error() {
+                    LanguageModelError::PermanentError(e.into())
+                } else if status.is_server_error() {
+                    LanguageModelError::TransientError(e.into())
```

Comment:
I'm not sure. What point is there in retrying if the remote server doesn't work? They had an outage this weekend, it would only have clobbered more resources on both ends than result in anything productive.

Reply:
That's what backoff is for. If the servers are clobbered, it backs off exponentially until the servers are no longer clobbered. Additionally, if there is a more long term problem, then backoff eventually gives up too.